// @clr/core/internal/utils/global.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { isBrowser } from './exists.js';
import { getAngularVersion, getReactVersion, getVueVersion, getAngularJSVersion } from './framework.js';
function getVersion() {
    const log = {
        versions: window.CDS._version,
        loadedElements: window.CDS._loadedElements,
        userAgent: navigator.userAgent,
        angularVersion: getAngularVersion(false),
        angularJSVersion: getAngularJSVersion(false),
        reactVersion: getReactVersion(false),
        vueVersion: getVueVersion(false),
    };
    return log;
}
function logVersion() {
    console.log(JSON.stringify(getVersion(), null, 2));
}
function initializeCDSGlobal() {
    window.CDS = window.CDS || {
        _version: [],
        _loadedElements: [],
        _react: { version: undefined },
        getVersion,
        logVersion,
    };
}
function setRunningVersion() {
    const loadedVersion = '4.0.5';
    if (window.CDS._version.indexOf(loadedVersion) < 0) {
        window.CDS._version.push(loadedVersion);
    }
    document.querySelector('body').setAttribute('cds-version', window.CDS._version[0]);
    if (window.CDS._version.length > 1) {
        console.warn('Running more than one version of Clarity can cause unexpected issues. Please ensure only one version is loaded.');
    }
}
export function setupCDSGlobal() {
    if (isBrowser()) {
        initializeCDSGlobal();
        setRunningVersion();
    }
}
