// @clr/core/forms/control-action/control-action.element.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { html, property } from 'lit-element';
import { baseStyles, CdsBaseButton, querySlot, setAttributes, assignSlotNames } from '@clr/core/internal';
import { styles } from './control-action.element.css.js';
import { LogService } from '@clr/core/internal';
/**
 * Control Action
 *
 * ```typescript
 * import '@clr/core/forms/register.js';
 * ```
 *
 * ```html
 * <cds-control-action>
 *
 * </cds-control-action>
 * ```
 * @internal
 * @element cds-control-action
 * @slot - For projecting text content or cds-icon
 */
export class CdsControlAction extends CdsBaseButton {
    static get styles() {
        return [baseStyles, styles];
    }
    render() {
        return html `
      <div class="private-host">
        <slot></slot>
      </div>
      ${this.hiddenButtonTemplate}
    `;
    }
    connectedCallback() {
        super.connectedCallback();
        setAttributes(this, ['aria-hidden', 'true']);
    }
    updated(props) {
        super.updated(props);
        if (props.has('action')) {
            this.setSlotLocation();
        }
        if (props.has('readonly')) {
            this.validateAriaLabel();
        }
    }
    setSlotLocation() {
        var _a;
        assignSlotNames([this, (_a = this.action) !== null && _a !== void 0 ? _a : false]);
    }
    validateAriaLabel() {
        var _a;
        if (!this.readonly && !((_a = this.getAttribute('aria-label')) === null || _a === void 0 ? void 0 : _a.length)) {
            LogService.warn('A aria-label is required for interactive cds-control-actions', this);
        }
    }
}
__decorate([
    property({ type: String })
], CdsControlAction.prototype, "action", void 0);
__decorate([
    querySlot('cds-icon')
], CdsControlAction.prototype, "icon", void 0);
